demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1540 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Demo for showing the use of MQTT APIs to establish an MQTT session,
+ * subscribe to a topic, publish to a topic, receive incoming publishes,
+ * unsubscribe from a topic and disconnect the MQTT session.
+ *
+ * A mutually authenticated TLS connection is used to connect to the broker
+ * in this example. Define SERVER_CERT_PATH, CLIENT_CERT_PATH,
+ * and CLIENT_PRIVATE_KEY_PATH in demo_config.h to achieve mutual
+ * authentication.
+ *
+ * The example is single threaded and uses statically allocated memory;
+ * it uses QOS1 and therefore implements a retransmission mechanism
+ * for Publish messages. Retransmission of publish messages are attempted
+ * when a MQTT connection is established with a session that was already
+ * present. All the outgoing publish messages waiting to receive PUBACK
+ * are resend in this demo. In order to support retransmission all the outgoing
+ * publishes are stored until a PUBACK is received.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX socket includes. */
+#include <netdb.h>
+#include <poll.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+/* MQTT API header. */
+#include "mqtt.h"
+
+/* Demo Config header. */
+#include "demo_config.h"
+
+
+/* Check that client identifier is defined. */
+#ifndef CLIENT_IDENTIFIER
+    #error "Please define a unique CLIENT_IDENTIFIER in demo_config.h."
```

Comment:
Since these are empty strings in the demo_config we will never get this error?

Reply:
My intention was to throw an error if this application code is copied over and try to build it before defining the macros. I have cross referenced the demos in v4_beta and a:FR and found that the approach taken is like this https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/demos/src/iot_demo_mqtt.c#L52-#L69. I will follow the same approach as it makes more sense.